src/Services/Implementations/ReverseGeocodes/GoogleMapsReverseGeocodeService.cs

```diff
@@ -28,13 +35,22 @@ public async Task<IEnumerable<string>> Get(Coordinate coordinate, string? langua
 		return addressPropertyValueList;
 	}
 
-	public async Task<GoogleMapsResponse?> SerializeFullResponse(Coordinate coordinate, string? language)
+	public async Task<GoogleMapsResponse?> SerializeFullResponse(ReverseGeocodeRequest request)
 	{
 		try
 		{
-			var queryString = $"?latlng={coordinate.Latitude},{coordinate.Longitude}&key={_apiKeyStore.GoogleMaps}";
-			if (language != null)
-				queryString += $"&language={language}";
+			var queryString = $"?latlng={request.Coordinate.Latitude},{request.Coordinate.Longitude}&key={_apiKeyStore.GoogleMaps}";
+
+			if (request.Language != null)
+			{
+				queryString += $"&language={request.Language}";
+			}
+
+			if (_cache.TryGetValue(request, out GoogleMapsResponse? cachedData))
+			{
+				return cachedData;
```

Comment:
This code is very straightforward easily testable by manual tests if the same coordinate bucket is requested from cache. It will be perfect to write an unit test to verify this cache hit logic.

This unit test will also cover your ReverseGeocodeRequest equality and it's hash code implementation.

Creating an new unit test on GoogleMapsCloudReverseGeocodeServiceUnitTests just to cover cache hit & miss scenarios would be perfect.

Reply:
I encountered difficulties testing the code in its current state due to the lack of support for `setup/verification` expressions in the `Dictionary `class, as it is not overridable. Attempting to use `IDictionary` resulted in implementation errors related to dependency injection. Ultimately, I resolved the issue by creating an `ICoordinateCache` interface. This not only addressed [Caching Reverse Geocode Result](https://github.com/photo-cli/photo-cli/issues/5) but also lays the groundwork for a potential evolution into an in-memory database in the future. I'd appreciate hearing your thoughts on this approach. @alpcoker

Reply:
I think that `ICoordinateCache` interface is more extensible for now and future usages comparing to Dictionary injection. Thanks for refactoring.

I checkout your code and refactored a little bit. The only problem I see that each reverse geocode provider may able to reach other provider cache. Also that `ICacheCoordinate` interface and it's implementations seems to me that can be easily make as generic. In the future `ICacheCoordinate` implementations could go from generic approach to separate implementations when we go into database implementation but interface can still be generic.

I am sharing my changes as patch. What do you think? If it's good for you also, can you squash this changes, @RecepCil ?

[generic_cache.patch](https://github.com/photo-cli/photo-cli/files/13777465/generic_cache.patch)

Reply:
I've implemented the changes you recommended, and I can already see how it has made the code more generic and easier to read. If you have any further feedback or if there's anything else you think needs attention, please feel free to let me know.

Reply:
Thanks Recep. I think the whole implementation is very good. 

One more small thing, can we use your new class while requesting on OpenStreetMap like I did on that patch ? 
[using_ReverseGeocodeRequest.patch](https://github.com/photo-cli/photo-cli/files/13854761/using_ReverseGeocodeRequest.patch)

Can you also squash all this changes into one commit ?